src/fastertransformer/models/llama/Llama.cc

```diff
@@ -0,0 +1,1236 @@
+    ftNcclStreamSynchronize(tensor_para_, pipeline_para_, stream_);
+}
+
+template<typename T>
+void Llama<T>::setOutputTensors(std::unordered_map<std::string, Tensor>*       output_tensors,
+                                  const std::unordered_map<std::string, Tensor>* input_tensors,
+                                  const size_t                                   max_input_length,
+                                  const size_t                                   max_output_seq_len)
+{
+    FT_LOG_DEBUG(__PRETTY_FUNCTION__);
+    if (pipeline_para_.rank_ != pipeline_para_.world_size_ - 1) {
+        return;
+    }
+
+    const size_t batch_size       = output_tensors->at("output_ids").shape[0];
+    const size_t beam_width       = output_tensors->at("output_ids").shape[1];
+    uint*        sequence_lengths = output_tensors->at("sequence_length").getPtr<uint>();
+    const size_t max_prefix_soft_prompt_length =
+        has_prefix_soft_prompt_ ? input_tensors->at("request_prompt_embedding").shape[1] : 0;
+
+    if (input_tensors->at("input_ids").shape[1] == 0) {
+        invokeCudaD2DcpyConvert(
+            sequence_lengths, sequence_lengths_, output_tensors->at("sequence_length").size(), stream_);
+        // TODO: D2D sequence_lenghts
+        if (beam_width > 1) {
+            // For beam search, do gather_tree
+            // take output_parent_ids as inter buffer
+            invokeGatherTree(transposed_output_ids_buf_,
+                             sequence_lengths_,
+                             max_output_seq_len,
+                             batch_size,
+                             beam_width,
+                             output_ids_buf_ + batch_size * beam_width,
+                             parent_ids_buf_ + batch_size * beam_width,
+                             end_ids_buf_,
+                             stream_);
+
+            // transpose and take output_parent_ids as inter buffer
+            invokeTransposeAxis01(output_tensors->at("output_ids").getPtr<int>(),
+                                  transposed_output_ids_buf_,
+                                  max_output_seq_len - 1,
+                                  batch_size * beam_width,
+                                  1,
+                                  stream_);
+        }
+        else {
+            // For sampling, only copy the results to output_tensor
+            invokeTransposeAxis01(output_tensors->at("output_ids").getPtr<int>(),
+                                  output_ids_buf_ + batch_size * beam_width,
+                                  max_output_seq_len - 1,
+                                  batch_size * beam_width,
+                                  1,
+                                  stream_);
+        }
+    }
+    else {
+
+        // For sampling, it is equivalent to all parent ids are 0.
+        gatherTreeParam param;
+        param.beams                = transposed_output_ids_buf_;
+        param.max_sequence_lengths = sequence_lengths_;
```

Comment:
```suggestion
        param.max_sequence_lengths = sequence_lengths;
```

Reply:
In general this function is quite similar to https://github.com/NVIDIA/FasterTransformer/blob/c6e8f60ec40da218804a60e6aa986903e7fa8594/src/fastertransformer/models/multi_gpu_gpt/ParallelGpt.cc#L1644-L1734 but slightly different. I wonder why we have such a divergence.

Reply:
Hi @yinghai , the code is just a quick copy&paste from `GptNeoX` model. To be honest, I don't know why we need that divergence either..

Reply:
Yeah that's what I figured. Unfortunately, the gptneo-x impl is wrong (https://github.com/NVIDIA/FasterTransformer/pull/589). :)

Reply:
merged the fix, thanks @yinghai !